### PR TITLE
Sensor config [WiP]

### PR DESCRIPTION
**Pull request by adamheinrich**
Not well tested yet :-)

This adds the following parameters:

- `operation_mode` (string, operation mode from *Table 3-3: Operating modes overview*)
- `acc_range` (int, values from *Table 3-8: Accelerometer configurations* [G])
- `acc_bandwidth` (double, values from *Table 3-8: Accelerometer configurations* [Hz])
- `gyro_range` (int, values from *Table 3-9: Gyroscope configurations* [dps])
- `gyro_bandwidth` (double, values from *Table 3-9: Gyroscope configurations* [Hz])

The `{acc,gyro}_{range,bandwidth}` parameters are configured only when `operation_mode` is not in the Fusion mode, otherwise the range and bandwidth are controlled by the fusion algorithm.

Example:

```
        <param name="operation_mode" type="string" value="ACCGYRO"/>
        <param name="acc_bandwidth" type="double" value="31.25"/>
```

**STEPS TO ENABLE BITBANGING ON  RASPBERRYPI:** 

IMU BNO055 is used with communication setup through I2C. If bitbanging is not enabled high random peaks in IMU raw and fused data can be experienced

First disable the HW I2C in `/boot/config.txt` -- just comment it out by putting `#` infront of the following line so in the end it looks like this:

    #dtparam=i2c_arm=on

Then add the following line to use a bit-banged driver for /dev/i2c-1 on the
"original" I2C pins:

    dtoverlay=i2c-gpio,i2c_gpio_sda=2,i2c_gpio_scl=3,bus=1

After reboot, check dmesg to verify the bit-banged driver is correctly initialized:

	ubuntu@ubiquityrobot:~$ dmesg | grep -i i2c
	[    2.882085] i2c-gpio 100000002.i2c: using lines 2 (SDA) and 3 (SCL)

### CODE REVIEW

**Comment by adamheinrich** [comment]
I merged my work in the atomic_read branch and also added parameters which allow to select what data is being read from the IMU:

```
        <param name="enable_raw" type="bool" value="true"/>
        <param name="enable_data" type="bool" value="false"/>
        <param name="enable_status" type="bool" value="false"/>
```

- `raw` are data published in `/imu/raw` and `/imu/mag`, 18 bytes in total
- `data` are the fused data published in `/imu/data`, 26 bytes in total
- `status` are some status registers and temperature published in `/imu/temperature`, 7 bytes in total

The change of the I2C read API merged from atomic_read should already improve CPU consumption a bit when used together with the bit-banged I2C driver (i2c-gpio).

**Comment by JanezCim** [comment]
I've ran it with 
```
    <node ns="imu" name="imu_node" pkg="imu_bno055" type="bno055_i2c_node" respawn="true" respawn_delay="2">
        <param name="device" type="string" value="/dev/i2c-1"/>
        <param name="address" type="int" value="40"/> <!-- 0x28 == 40 is the default for BNO055 -->
        <param name="frame_id" type="string" value="imu"/>
        <param name="operation_mode" type="string" value="ACCGYRO"/>
        <param name="acc_bandwidth" type="double" value="31.25"/>
        <param name="enable_raw" type="bool" value="true"/>
        <param name="enable_data" type="bool" value="false"/>
        <param name="enable_status" type="bool" value="false"/>
    </node>
```
and its reaching 42% usage of one CPU which is better then before (~70%) but still quite large. I think we can also have a param to limit the frequency? We dont need 100Hz data anyway. But the spikes are totaly gone with the bitbanger and it seems like the bandwidth is helping somewhat with filtering the vibrations of the PID on the chassis

image
(on this graph i am first teleoping the robot to go straight forward and then spin on the spot)

**Comment by adamheinrich** [comment]
The frequency can be already conrolled (param `rate`): https://github.com/JanezCim/ros-imu-bno055/blob/master/imu_bno055/src/bno055_i2c_node.cpp#L59

**Comment by JanezCim** [comment]
You are right, at rate=30, we are down to 12%, which I think is still large, but acceptable. Now I am seeing zero spikes, which is great. Now we only have the issue that the gravity is not compensated in raw data. Will try out some ROS solutions for that. Now here is the chassy vibration readings with and without power to the motors. So while the data is much nicer then what we started with, changing the bandwidth does not produce a huge difference in the chassis vibration reading reduction.

acc_bandwidth=7.81
image

acc_bandwidth=31.25
image

acc_bandwidth=125.0
image

acc_bandwidth=500.0
image

**Comment by adamheinrich** [comment]
Datashet:

> The time difference between the successive acceleration samples depends on the selected bandwidth and equates to 1/(2 * bw)

For bandwidth of 125Hz, the IMU is reading accelerometer at 250Hz.

**Comment by JanezCim** [comment]
Another set of tests:

```
<param name="acc_bandwidth" type="double" value="15.63"/>
<param name="rate" type="int" value="32"/>
```
image


and 
```
<param name="acc_bandwidth" type="double" value="1000"/>
<param name="rate" type="int" value="32"/>
```
image

**Comment by JanezCim** [comment]
Another set of tests:

```
<param name="acc_bandwidth" type="double" value="15.63"/>
<param name="rate" type="int" value="32"/>
```
image


and 
```
<param name="acc_bandwidth" type="double" value="1000"/>
<param name="rate" type="int" value="32"/>
```
image

Now we are at 14% CPU usage. Well those are much better results!

**Comment by JanezCim** [comment]
Now the robot going forward and then turning on spot tests:

acc_bandwidth=7.81
image

acc_bandwidth=15.63
image

acc_bandwidth=31.25
image

acc_bandwidth=62.5
image

acc_bandwidth= 125.0
image

**Comment by JanezCim** [comment]
seems alright so far, im going to go with that, I think for now the only addition I would make to the repo is that the topics arent even initialized if I disable them. Im currently running:
```
        <param name="enable_raw" type="bool" value="true"/>
        <param name="enable_data" type="bool" value="false"/>
        <param name="enable_status" type="bool" value="false"/>
```

and what i see is: 
```
ubuntu@ubiquityrobot ~ $ rostopic list | grep imu
/imu/data
/imu/mag
/imu/raw
/imu/status
/imu/temp
```

what i expect to be seein is: 
```
ubuntu@ubiquityrobot ~ $ rostopic list | grep imu
/imu/raw
/imu/temp
```

File: imu_bno055/include/imu_bno055/bno055_i2c_activity.h [hunk 1]
```diff
@@ -253,7 +258,15 @@ class BNO055I2CActivity {
     // ROS parameters
     std::string param_frame_id;
     std::string param_device;
+    std::string param_operation_mode;
     int param_address;
+    double param_acc_bandwidth;
```
Comment (mjstn) [inline imu_bno055/include/imu_bno055/bno055_i2c_activity.h:263]: These must be what you were discussing in meeting.

File: imu_bno055/src/bno055_i2c_activity.cpp [hunk 2]
```diff
@@ -55,9 +64,119 @@ BNO055I2CActivity::BNO055I2CActivity(ros::NodeHandle &_nh, ros::NodeHandle &_nh_
 
 // ******** private methods ******** //
 
+int BNO055I2CActivity::operation_mode() {
+    std::map<std::string, int> mode_map {
+        { "ACCONLY", BNO055_OPERATION_MODE_ACCONLY },
```
Comment (mjstn) [inline imu_bno055/src/bno055_i2c_activity.cpp:69]: string names will be useful

**Review by mjstn — APPROVED** [review]
Seems to be in order. Looks good.

**Comment by adamheinrich** [comment]
@JanezCim:

> seems alright so far, im going to go with that, I think for now the only addition I would make to the repo is that the topics arent even initialized if I disable them.

Unfortunately I don't immediately see how to achieve this. What I did is that I execute e.g. `pub_data.publish()` conditionally when `param_enable_data` is nonzero ([link](https://github.com/JanezCim/ros-imu-bno055/pull/2/commits/c8c0faecfc830142eb2c2ca024e2d91cb7b0c80b#diff-d123d33953218f77165d23e895031eedf7fa256f0e4cfe9c7e648c2ee68d0401L356-R439)), I did not touch other code so the part where topics are advertised stayed as it was ([link](https://github.com/JanezCim/ros-imu-bno055/blob/sensor_config/imu_bno055/src/bno055_i2c_activity.cpp#L223)).

There is not much more available in the [ros::Publisher API](http://docs.ros.org/en/indigo/api/roscpp/html/classros_1_1Publisher.html#a580af4f9f7ab80876036d8aa25dd27d5) than `publish()` anyway.

EDIT: I'm an idiot, you wanted to **not** have the topics initialized of course. Need to read twice next time :)

**Comment by adamheinrich** [comment]
I think parts of this are worth contributing upstream after a small cleanup. Probably as three PRs so the maintainer can select what is useful: atomic read, range/bw configuraton and configurable read groups.

**Comment by JanezCim** [comment]
Hey, ive pushed a suggestion oh now to  enable/disable also ros topics, can you take a look. This works fine with the robot

**Review by JanezCim — APPROVED** [review]
I am now fine with merging

**Pull request merged**